File: legacy_python/backend/matrix.py
"""Phase 2 — 3-way combinatorial matrix generation.

Implementation notes / improvements over the spec:
  * Naive enumeration of all 3-way combinations explodes quickly. We cap the factor
    level count at MAX_LEVELS_PER_FACTOR and then use a *greedy maximum-coverage*
    selection (effectively a randomized greedy approximation of IPOG-F).
  * We always include `persona` and `objective` as fixed factors in every test row
    (since they directly drive the tester agent), then fill the remaining factors
    by maximizing 3-way coverage.
  * If the user supplies many factors, we limit the total rows to `max_rows` to keep
    the test-suite tractable; the coverage % is reported.
"""
from __future__ import annotations

import itertools
import random
from typing import Any, Iterable

from .models import Factor, TestConfiguration, TestMatrix, TestVariables
from .parsing import level_value


MAX_LEVELS_PER_FACTOR = 6  # additional safety net


def _truncate_levels(factors: list[Factor]) -> list[Factor]:
    out = []
    for f in factors:
        if len(f.levels) > MAX_LEVELS_PER_FACTOR:
            f = f.model_copy(update={"levels": f.levels[:MAX_LEVELS_PER_FACTOR]})
        out.append(f)
    return out


def _all_triplets(factors: list[Factor]) -> set[tuple]:
    """All (factor_a, level_a, factor_b, level_b, factor_c, level_c) triplets."""
    triplets: set[tuple] = set()
    for fa, fb, fc in itertools.combinations(range(len(factors)), 3):
        for la, lb, lc in itertools.product(
            factors[fa].levels, factors[fb].levels, factors[fc].levels
        ):
            triplets.add(
                (
                    factors[fa].name,
                    _hashable(la),
                    factors[fb].name,
                    _hashable(lb),
                    factors[fc].name,
                    _hashable(lc),
                )
            )
    return triplets


def _hashable(v: Any) -> Any:
    if isinstance(v, dict):
        return tuple(sorted((k, _hashable(val)) for k, val in v.items()))
    if isinstance(v, list):
        return tuple(_hashable(x) for x in v)
    return v


def _coverage_of(assignment: dict[str, Any], factors: list[Factor]) -> set[tuple]:
    names = list(assignment.keys())
    covered: set[tuple] = set()
    for a, b, c in itertools.combinations(range(len(names)), 3):
        covered.add(
            (
                names[a],
                _hashable(assignment[names[a]]),
                names[b],
                _hashable(assignment[names[b]]),
                names[c],
                _hashable(assignment[names[c]]),
            )
        )
    return covered


def _random_assignment(factors: list[Factor], rng: random.Random) -> dict[str, Any]:
    return {f.name: rng.choice(f.levels) for f in factors}


def generate_3way_matrix(
    variables: TestVariables,
    max_rows: int = 30,
    candidates_per_iter: int = 40,
    seed: int = 1234,
) -> TestMatrix:
    """Greedy maximum-coverage 3-way pairwise generator.

    Each iteration we sample `candidates_per_iter` random configurations and pick
    the one that covers the most still-uncovered triplets.
    """
    rng = random.Random(seed)
    factors = _truncate_levels(variables.factors)

    if len(factors) < 3:
        # Fall back to full Cartesian product when there aren't enough dims.
        rows = []
        for combo in itertools.product(*[f.levels for f in factors]):
            rows.append(
                TestConfiguration(
                    assignments={factors[i].name: combo[i] for i in range(len(factors))}
                )
            )
            if len(rows) >= max_rows:
                break
        return TestMatrix(
            factors=factors, rows=rows, total_triplets=0, coverage_percent=100.0
        )

    remaining = _all_triplets(factors)
    total = len(remaining)
    rows: list[TestConfiguration] = []

    while remaining and len(rows) < max_rows:
        best_assignment: dict[str, Any] | None = None
        best_new = -1
        for _ in range(candidates_per_iter):
            assignment = _random_assignment(factors, rng)
            new = len(_coverage_of(assignment, factors) & remaining)
            if new > best_new:
                best_new = new
                best_assignment = assignment
                if new == 0:
                    continue
        if best_assignment is None or best_new <= 0:
            break

        covered = _coverage_of(best_assignment, factors) & remaining
        rows.append(
            TestConfiguration(
                assignments=best_assignment, covered_triplets=len(covered)
            )
        )
        remaining -= covered

    coverage_pct = 0.0 if total == 0 else 100.0 * (total - len(remaining)) / total
    return TestMatrix(
        factors=factors,
        rows=rows,
        total_triplets=total,
        coverage_percent=round(coverage_pct, 2),
    )


def assignment_to_db_config(
    assignment: dict[str, Any], factors: list[Factor]
) -> dict[str, Any]:
    """Project a matrix row down to just the DB state config the sandbox needs."""
    db_config: dict[str, Any] = {}
    for f in factors:
        if f.kind == "db_var" and f.name in assignment:
            db_config[f.source or f.name] = level_value(assignment[f.name])
    return db_config


def assignment_persona(assignment: dict[str, Any]) -> str:
    return str(assignment.get("persona", "Generic User"))


def assignment_objective(assignment: dict[str, Any]) -> str:
    return str(assignment.get("objective", "find any failure mode"))


def assignment_tool_hints(
    assignment: dict[str, Any], factors: Iterable[Factor]
) -> dict[str, Any]:
    """Tool-parameter hints we can pass to the tester to bias prompt generation."""
    hints: dict[str, Any] = {}
    for f in factors:
        if f.kind == "tool_param" and f.name in assignment:
            hints[f.source or f.name] = level_value(assignment[f.name])
    return hints

File: core_engine/matrix.ts
/**
 * Phase 2 of the test pipeline — greedy maximum-coverage 3-way combinatorial
 * matrix generation.
 *
 * Strict port of `legacy_python/backend/matrix.py`. The randomized greedy
 * approximation of IPOG-F is good enough for v1; tighter algorithms (proper
 * IPOG-F, all-pairs reduction) wait for Phase 5.
 */
import { randomUUID } from 'node:crypto';
import {
  Factor,
  FactorLevel,
  TestConfiguration,
  TestMatrix,
  TestVariables
} from './engine_types';

const MAX_LEVELS_PER_FACTOR = 6;

class SeededRandom {
  private state: number;
  constructor(seed: number) {
    this.state = seed >>> 0 || 1;
  }
  next(): number {
    // xorshift32 — fine for a deterministic test-case generator
    let x = this.state;
    x ^= x << 13;
    x ^= x >>> 17;
    x ^= x << 5;
    this.state = x >>> 0;
    return this.state / 0xffffffff;
  }
  pick<T>(arr: T[]): T {
    return arr[Math.floor(this.next() * arr.length) % arr.length];
  }
}

function truncateLevels(factors: Factor[]): Factor[] {
  return factors.map((f) =>
    f.levels.length > MAX_LEVELS_PER_FACTOR
      ? { ...f, levels: f.levels.slice(0, MAX_LEVELS_PER_FACTOR) }
      : f
  );
}

function levelKey(level: FactorLevel): string {
  if (level && typeof level === 'object' && 'value' in level) {
    return `o:${JSON.stringify(level)}`;
  }
  return `s:${JSON.stringify(level)}`;
}

/** Triplet representation: stable string for set membership. */
function tripletKey(
  fa: string,
  la: FactorLevel,
  fb: string,
  lb: FactorLevel,
  fc: string,
  lc: FactorLevel
): string {
  return `${fa}=${levelKey(la)}|${fb}=${levelKey(lb)}|${fc}=${levelKey(lc)}`;
}

function* triples(n: number): Generator<[number, number, number]> {
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      for (let k = j + 1; k < n; k++) {
        yield [i, j, k];
      }
    }
  }
}

function allTriplets(factors: Factor[]): Set<string> {
  const out = new Set<string>();
  for (const [i, j, k] of triples(factors.length)) {
    for (const la of factors[i].levels) {
      for (const lb of factors[j].levels) {
        for (const lc of factors[k].levels) {
          out.add(tripletKey(factors[i].name, la, factors[j].name, lb, factors[k].name, lc));
        }
      }
    }
  }
  return out;
}

function coverageOf(
  assignment: Record<string, FactorLevel>,
  factors: Factor[]
): Set<string> {
  const out = new Set<string>();
  const names = factors.map((f) => f.name);
  for (const [i, j, k] of triples(names.length)) {
    out.add(
      tripletKey(
        names[i],
        assignment[names[i]],
        names[j],
        assignment[names[j]],
        names[k],
        assignment[names[k]]
      )
    );
  }
  return out;
}

function randomAssignment(
  factors: Factor[],
  rng: SeededRandom
): Record<string, FactorLevel> {
  const out: Record<string, FactorLevel> = {};
  for (const f of factors) {
    out[f.name] = rng.pick(f.levels);
  }
  return out;
}

export interface MatrixOptions {
  maxRows?: number;
  candidatesPerIter?: number;
  seed?: number;
}

export function generate3WayMatrix(
  variables: TestVariables,
  opts: MatrixOptions = {}
): TestMatrix {
  const maxRows = opts.maxRows ?? 30;
  const candidatesPerIter = opts.candidatesPerIter ?? 40;
  const seed = opts.seed ?? 1234;

  const rng = new SeededRandom(seed);
  const factors = truncateLevels(variables.factors);

  // Less than 3 factors → just enumerate the cartesian product up to maxRows.
  if (factors.length < 3) {
    const rows: TestConfiguration[] = [];
    const buf: FactorLevel[] = new Array(factors.length);
    function* product(idx: number): Generator<FactorLevel[]> {
      if (idx === factors.length) {
        yield buf.slice();
        return;
      }
      for (const lvl of factors[idx].levels) {
        buf[idx] = lvl;
        yield* product(idx + 1);
      }
    }
    for (const combo of product(0)) {
      const assignments: Record<string, FactorLevel> = {};
      factors.forEach((f, i) => (assignments[f.name] = combo[i]));
      rows.push({ id: randomUUID().slice(0, 8), assignments, coveredTriplets: 0 });
      if (rows.length >= maxRows) break;
    }
    return { factors, rows, totalTriplets: 0, coveragePercent: 100.0 };
  }

  const remaining = allTriplets(factors);
  const total = remaining.size;
  const rows: TestConfiguration[] = [];

  while (remaining.size > 0 && rows.length < maxRows) {
    let bestAssignment: Record<string, FactorLevel> | null = null;
    let bestNew = -1;
    for (let i = 0; i < candidatesPerIter; i++) {
      const candidate = randomAssignment(factors, rng);
      const cov = coverageOf(candidate, factors);
      let newCount = 0;
      for (const k of cov) if (remaining.has(k)) newCount++;
      if (newCount > bestNew) {
        bestNew = newCount;
        bestAssignment = candidate;
      }
    }
    if (!bestAssignment || bestNew <= 0) break;

    const cov = coverageOf(bestAssignment, factors);
    let covered = 0;
    for (const k of cov) {
      if (remaining.delete(k)) covered++;
    }
    rows.push({
      id: randomUUID().slice(0, 8),
      assignments: bestAssignment,
      coveredTriplets: covered
    });
  }

  const coveragePct = total === 0 ? 0 : (100 * (total - remaining.size)) / total;
  return {
    factors,
    rows,
    totalTriplets: total,
    coveragePercent: Math.round(coveragePct * 100) / 100
  };
}

// ---------------------------------------------------------------------------
// row helpers used downstream
// ---------------------------------------------------------------------------

export function assignmentPersona(assignment: Record<string, FactorLevel>): string {
  const v = assignment['persona'];
  return typeof v === 'string' ? v : 'Generic User';
}

export function assignmentObjective(assignment: Record<string, FactorLevel>): string {
  const v = assignment['objective'];
  return typeof v === 'string' ? v : 'find any failure mode';
}

export function assignmentToolHints(
  assignment: Record<string, FactorLevel>,
  factors: Factor[]
): Record<string, any> {
  const out: Record<string, any> = {};
  for (const f of factors) {
    if (f.kind === 'tool_param' && f.name in assignment) {
      out[f.source || f.name] = unwrapLevel(assignment[f.name]);
    }
  }
  return out;
}

function unwrapLevel(level: FactorLevel): any {
  if (level && typeof level === 'object' && 'value' in level) return level.value;
  return level;
}
